python/dgl/readout.py

```diff
@@ -542,33 +542,42 @@ def _topk_on(graph, typestr, feat, k, descending, sortby, ntype_or_etype):
 
     length = max(max(F.asnumpy(batch_num_objs)), k)
     fill_val = -float('inf') if descending else float('inf')
-    feat_ = F.pad_packed_tensor(feat, batch_num_objs, fill_val, l_min=k)
-
-    if sortby is not None:
-        keys = F.squeeze(F.slice_axis(feat_, -1, sortby, sortby+1), -1)
-        order = F.argsort(keys, -1, descending=descending)
-    else:
-        order = F.argsort(feat_, 1, descending=descending)
-
-    topk_indices = F.slice_axis(order, 1, 0, k)
-
-    # zero padding
-    feat_ = F.pad_packed_tensor(feat, batch_num_objs, 0, l_min=k)
-
-    if sortby is not None:
-        feat_ = F.reshape(feat_, (batch_size * length, -1))
-        shift = F.repeat(F.arange(0, batch_size) * length, k, -1)
-        shift = F.copy_to(shift, F.context(feat))
-        topk_indices_ = F.reshape(topk_indices, (-1,)) + shift
+    feat_ = F.pad_packed_tensor(feat, batch_num_objs, fill_val, l_min=k)  # (batch_size, l, d)
+    if F.backend_name == 'pytorch' and sortby is not None:
+        import torch as th
+        keys = feat_[..., sortby]  # (batch_size, l)
+        topk_indices = keys.topk(k, -1, largest=descending)[1]  # (batch_size, k)
+        feat_ = F.pad_packed_tensor(feat, batch_num_objs, 0, l_min=k)
+        feat_ = feat_.view((batch_size * length), -1)
+        shift = th.arange(0, batch_size, device=feat_.device).view(batch_size, 1) * length
+        topk_indices_ = topk_indices + shift
+        return feat_[topk_indices_].view(batch_size, k, -1), topk_indices
```

Comment:
put the two branches into two functions `torch_topk` and `general_topk`.

Reply:
i see